sql/core/src/main/scala/org/apache/spark/sql/execution/reuse/Reuse.scala

```diff
@@ -0,0 +1,90 @@
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.reuse
+
+import scala.collection.mutable.Map
+
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.execution.{BaseSubqueryExec, ExecSubqueryExpression, ReusedSubqueryExec, SparkPlan}
+import org.apache.spark.sql.execution.exchange.{Exchange, ReusedExchangeExec}
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.types.StructType
+
+/**
+ * Find out duplicated exchanges and subqueries in the whole spark plan including subqueries, then
+ * use the same exhange or subquery for all the references.
+ */
+case class ReuseExchangeAndSubquery(conf: SQLConf) extends Rule[SparkPlan] {
+
+  private class ReuseCache[T <: SparkPlan] {
+    // To avoid costly canonicalization of an exchange or a subquery:
+    // - we use its schema first to check if it can be replaced to a reused one at all
+    // - we insert it into the map of canonicalized plans only when at least 2 have the same schema
+    private val cache = Map[StructType, (T, Map[SparkPlan, T])]()
+
+    def lookup(plan: T): T = {
+      val (firstSameSchemaPlan, sameResultPlans) = cache.getOrElseUpdate(plan.schema, plan -> Map())
+      if (firstSameSchemaPlan.ne(plan)) {
+        if (sameResultPlans.isEmpty) {
+          sameResultPlans += firstSameSchemaPlan.canonicalized -> firstSameSchemaPlan
+        }
+        sameResultPlans.getOrElseUpdate(plan.canonicalized, plan)
+      } else {
+        plan
+      }
+    }
+  }
+
+  def apply(plan: SparkPlan): SparkPlan = {
+    if (conf.exchangeReuseEnabled || conf.subqueryReuseEnabled) {
+      val exchanges = new ReuseCache[Exchange]()
+      val subqueries = new ReuseCache[BaseSubqueryExec]()
+
+      def reuse(plan: SparkPlan): SparkPlan = plan.transformUp {
+        case exchange: Exchange if conf.exchangeReuseEnabled =>
+          val cached = exchanges.lookup(exchange)
+          if (cached.ne(exchange)) {
+            ReusedExchangeExec(exchange.output, cached)
+          } else {
+            exchange
+          }
+
+        case other => other.transformExpressionsUp {
+          case sub: ExecSubqueryExpression =>
+            val subquery = reuse(sub.plan).asInstanceOf[BaseSubqueryExec]
+            sub.withNewPlan(
+              if (conf.subqueryReuseEnabled) {
```

Comment:
I like the new abstraction for hiding the loop, how about one step further and also abstract over this pattern when the found element is the same instance as the one in the cache, something like this:

```
def lookup(plan: T, f: T => T) {
	val res = ....
	if (res eq plan) {
		plan
	} else {
		f(res)
	}
}
```

and then the call site becomes: `lookupOrElse(subquery, ReusedSubqueryExec(_))`

Reply:
The name `lookupOrElse` sounds a bit weird to me, it suggests that `f` is applied if lookup fails (item isn't in the map yet). But in this case `f` should be applied if lookup founds the item in the map. So if we want that abstraction, shouldn't we call the method `putOrElse` or `addOrElse`?

Reply:
I added `addOrElse` in the latest commits.

Reply:
Yeah, I agree `lookupOrElse` is not the best name. I was thinking more about it, so we want a method with the following semantics:

Try  to find an existing node with the same canonicalized form:
- if no such node found, add the node to the map and return the node itself
- if a node found that refers the same instance as the key, do nothing, and return the node itself
- otherwise, call the given function on the retrieved node from the map and then return the result

I think it's not like anything (I mean common methods) on the Map interfaces that I know of. Somehow reminds me of Java 8 `putIfAbsent` method, but with the difference it calls the passed function if it's present. 

On second thought, maybe we should rename the `lookup` method to `getOrElseUpdate` because it doesn't just lookup. 
And call the second method `applyIfPresent`. Maybe also doesn't make sense to have the second method anymore since it's very specific and hard to capture with a method name what's going on.

One last thing, what was happening before if we had two exchanges in the query plan that were referring to the same instance? Were we leaving them intact or replacing it with a reuse node? Because now we just leave it as is. I'm not sure though if such a situation actually can happen to have to exchanges referring to the same instance in the query plan, if not, maybe we can remove the check and (the second method) altogether. Also maybe putting a reuse node when there are the same instances is even better, it's just a wrapper and signals the presence of the same exchange node.

Reply:
I agree that the name `lookup` doesn't capture exactly what the method does, but `getOrElseUpdate` has well known meaning and parameter list on scala maps and the first method doesn't match to that so I would use a different name like `getOrElseAdd` or `lookupOrElseAdd`.
Yes, it's even harder to find a good name for the second method. How about `reuseOrElseAdd`? It doesn't match any of the existing methods of a map and captures both reuse (with applying `f`) and add functionality.

The old behaviour was that on the second encounter of the same instance it was wrapped into a reuse node. This actually works different in this PR and the instance doesn't get wrapped on the second encounter. I did this way for 3 reasons:
- From performance perspective it doesn't make any difference
- An exchange or subquery instance can appear 2 times in the plan only if someone manually crafted the plan, which I don't think happens in a real word use case
- I'm using `getOrElseUpdate` 2 times in `lookup` to simplify the code, but from the result of `getOrElseUpdate` we can't tell if the key had already been in the map or it was added during the call. Besides simple code I also like using `getOrElseUpdate` because some concurrent map implementation supports it as atomic operation. I admit it doesn't matter in this PR, but I think it would be nice to incorporate `ReuseAdaptiveSubquery` functionality into this rule in a separate PR in the future.

Reply:
`reuseOrElseAdd` is probably a better name, but this naming is getting hard :-) About the new behavior, sounds good to me, but probably if there is no performance benefit and it's only happening in the tests, I would have just wrap the same instance in a reuse node to also not bother about the second method and the equality check, but don't have a strong opinion on this.